### PR TITLE
Add CodeQL workflow

### PR DESCRIPTION
**Pull request by ryao**
### Motivation and Context
CodeQL is a static analyzer from github with a very low false positive rate. We have long wanted to have static analysis runs done on every pull request and using CodeQL, we can.

### Description
We add a CodeQL workflow.

### How Has This Been Tested?
I tried a version of this in a branch on my github zfs repository.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by andrewc12** [comment]
Is there a specific reason this is limited to push or pull_request on master? None of the other workflows are setup like this, and it wont automatically run if you have a fork and dont use master.

**Comment by ryao** [comment]
> Is there a specific reason this is limited to push or pull_request on master? None of the other workflows are setup like this, and it wont automatically run if you have a fork and dont use master.

The reason is inexperience with github workflows. I had considered doing other branches, but I did not know if I could use a wild card operator. I heard from @behlendorf that this will be blocked until the issues that it highlighted are fixed so it is merged with a clean repository so there is time to revise this. I plan to open a PR with fixes to them relatively soon.

Also, I wonder how this would interact with PRs on branches when set to run on multiple branches in the master branch when apparently, it looks at the PR to see if it should run the git workflow, so what we have in the master branch would seem like it should not matter, although that needs to be checked.

**Comment by ryao** [comment]
I probably should add that this was put into a PR slightly prematurely. I had spent the OpenZFS developer summit hackathon mostly on code review of other people's code. The hackathon had prizes for the best work, and since code review is not tangible, I did a revision to what I had and put it into a PR to enter the competition. However, it was less impressive than a number of other things, so using it to enter the hackathon's best project competiton was largely in vain. Perhaps it was worth some nerd credit, which I would consider to make my stunt in putting this into a PR during the hackathon worthwhile, but that is hard to quantify.

**Comment by andrewc12** [comment]
I'm only slightly complaining, mostly in a "hey don't hoard the shiny things" sort of way.
But it wont actually end up affecting me since I've done it myself.
https://github.com/andrewc12/openzfs/tree/andrew_codeql-03

**Comment by ryao** [comment]
@andrewc12 There is room for both Linux and Windows CodeQL workflows. It is a shame that it does not support FreeBSD (although I guess we could open an issue in the codeql github repository requesting it).

Coincidentally, I had just pinged @lundman on the OpenZFS slack server about adding a workflow for Microsoft's C++ Code Analysis in the openzfsonwindows/openzfs repository a little while before you commented here. If you are interested in this, you could make a workflow for that. It requires CMake, which I believe has been added to avoid expecting Linux/FreeBSD/etcetera developers to modify visual studio project files.

**Comment by andrewc12** [comment]
> There is room for both Linux and Windows CodeQL workflows.

Cool, good to know.

> Coincidentally, I had just pinged @lundman on the OpenZFS slack server about adding a workflow for Microsoft's C++ Code Analysis in the openzfsonwindows/openzfs repository a little while before you commented here. If you are interested in this, you could make a workflow for that. It requires CMake, which I believe has been added to avoid expecting Linux/FreeBSD/etcetera developers to modify visual studio project files.

Thanks for the info I'll definitely done that if I can.

**Comment by ryao** [comment]
> > There is room for both Linux and Windows CodeQL workflows.
> 
> Cool, good to know.

As I explained in github/codeql#10982, having CodeQL for each platform enables us to scan the platform specific code, which would be missed if we only ran it on a single platform.

File: .github/workflows/codeql.yml
Comment (gmelikov) [inline .github/workflows/codeql.yml:56]: Wow, it really did something!

**Review by gmelikov — CHANGES_REQUESTED** [review]
It's really well, thank you @ryao !

We should fix it's warnings before merge, of course.

File: .github/workflows/codeql.yml
Comment (gmelikov) [reply .github/workflows/codeql.yml:56]: I'm not sure that python checks can find our files without `.py` extension without manual build, though. But it's still good for a start.

File: .github/workflows/codeql.yml
Comment (andrewc12) [reply .github/workflows/codeql.yml:16]: Yes please remove this restriction

**Comment by behlendorf** [comment]
This is nice.  As already mentioned above once we can wrap up the outstanding detects we can merge this.

File: .github/workflows/codeql.yml
Comment (ryao) [reply .github/workflows/codeql.yml]: I just pushed a revision intended to do this. I based on it zfs-tests-functional.yml.

File: .github/workflows/codeql.yml
Comment (ryao) [reply .github/workflows/codeql.yml:30]: I am not sure either. I should try putting a bug into the python code in a branch somewhere to test it, but I am not sure what python bugs it is designed to catch offhand and my python knowledge is very limited. I will leave this as something to check after the C warnings are resolved.

File: .github/workflows/codeql.yml
Comment (ryao) [reply .github/workflows/codeql.yml]: I just removed them and repushed.

File: .github/workflows/codeql.yml
Comment (ryao) [reply .github/workflows/codeql.yml]: Done.

**Comment by andrewc12** [comment]
@behlendorf from what I understand this workflow/check does not fail/block

As shown by this example of it passing.

chrome_screenshot_1666828037214

Instead it puts them as warnings in github itself as shown by this example
chrome_screenshot_1666827937615

So you don't actually need to fix the defects before mergeing this.

**Comment by behlendorf** [comment]
Good point, it looks like this is configurable in the repository so we could initially at least set a lower threshold.

https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#defining-the-severities-causing-pull-request-check-failure

**Comment by gmelikov** [comment]
Hm, any warning should fail check, I've missed that.

**Comment by ryao** [comment]
@behlendorf It turns out that there is one false positive:

https://github.com/ryao/zfs/security/code-scanning/19

CodeQL correctly points out that there is a race if you open a file, close it and open it again, but doing that with a different suid is the entire point of this test. I suspect that I could use `fork()` in a creative way to get around it, but my feeling is that we are better off leveraging github's ability to mark false positives. Presumably, if we do that, we will not get any more reports about it.

Also, since it says "22 new alerts including 21 high severity security vulnerabilities", it would seem that it only reports new issues in PRs, so even if we do not mark it as a false positive, we probably will not hear about it more than once. I suspect that the reason we hear about it here is because since because this is the PR for adding the code scans, every issue is new.

**Comment by gmelikov** [comment]
@behlendorf could you please lower checks for openzfs-docs too, while at it?

**Review by behlendorf — APPROVED** [review]

**Comment by behlendorf** [comment]
Since we can control if this reports as a CI failure I'm going to go ahead and merge this, we'll see how it goes.  I suppose one nice thing is we'll be able to test out things like marking false positives as noted by @ryao.

@gmelikov I'm not sure I follow.  We haven't enabled CodeQL for openzfs-docs so there's nothing to set yet.  But if/when we do I'm happy to change it.

**Comment by ryao** [comment]
@behlendorf That sounds good to me.

**Review by gmelikov — COMMENTED** [review]
LGTM

**Pull request merged**
